lib/game.rb

```diff
@@ -0,0 +1,167 @@
+##
+class Game
+  # => creates getters and setters for the instance variables
+  attr_accessor :board, :player_1, :player_2, :num_players
+
+  ##
+  # => Instantiates a game, defaults are two players and an empty board
+  ##
+  def initialize(player_1 = Players::Human.new("X"), player_2 = Players::Human.new("O"), board = Board.new)
+    @player_1 = player_1;
+    @player_2 = player_2;
+    @board = board;
+  end
+
+  # => All of the possible winning combinations
+  WIN_COMBINATIONS = [
+    [0, 1, 2], # Top row
+    [3, 4, 5], # Middle row
+    [6, 7, 8], # Bottom row
+    [0, 3, 6], # First column
+    [1, 4, 7], # Second column
+    [2, 5, 8], # Third Column
+    [0, 4, 8], # Left to Right diagonal
+    [2, 4, 6]  # Right to Left diagonal
+  ];
+
+  ##
+  # => Asks how many human players and sets the players up accordingly
+  ##
+  def define_players
+    puts "How many players want to play? (0, 1, or 2)"
+    @num_players = gets.strip.to_i
+
+    get_player_input = "Would you like to be 'X's or 'O's?";
+    #binding.pry
+    case @num_players
+      when 0
+          @player_1 = Players::Computer.new("X")
+          @player_2 = Players::Computer.new("O")
+      when 1
+        puts get_player_input;
+        input = gets.strip;
+
+        input == "O" || input == "o" ? @player_1.token = "O" && @player_2 = Players::Computer.new("X") : @player_2 = Players::Computer.new("O")
+      when 2
+        puts get_player_input;
+        input = gets.strip;
+        if input == "O" || input == "o"
+          @player_1.token = "O" && @player_2.token = "X"
+        end
+      else
+        puts "Invalid input, please select a valid option.";
+        define_players;
+      end
+  end #end #define_players
+
+  ##
+  # => This figures out who's turn it is
+  ##
+  def current_player
+    @board.turn_count % 2 == 0 ? @player_1 : @player_2;
+  end
+
+  ##
+  # => This checks if somebody has won the game or not
+  ##
+  def won?
+    x = "X";
+    o = "O";
+
+    WIN_COMBINATIONS.each do |win_combination|
+      win_index_1 = win_combination[0];
+      win_index_2 = win_combination[1];
+      win_index_3 = win_combination[2];
+
+      position_1 = @board.cells[win_index_1];
+      position_2 = @board.cells[win_index_2];
+      position_3 = @board.cells[win_index_3];
+
+      if ((position_1 == x && position_2 == x && position_3 == x) ||
+        (position_1 == o && position_2 == o && position_3 == o))
+          return win_combination;
+        else
+          false;
+      end
+    end
+    false; #explicitly tell ruby to return false if we've cycled through the board and no win combinations can be found
+  end
+
+  ##
+  # => This checks if there is a draw or not
+  ##
+  def draw?
+    @board.full? && !won?;
+  end
+
+  ##
+  # => This checks if the game has finished in a win or a draw
+  ##
+  def over?
+    won? || draw?;
+  end
+
+  ##
+  # => This determines who the winner is based on the winning combo
+  ##
+  def winner
```

Comment:
Here is a potential refactor for this method:
```ruby
if won = won?
  board.cells[won.first]
end
```

If `won?` returns true, then and only then will the local variable `won` be assigned, otherwise, the method will just abort with a false return value. But if `won` is assigned, we can use that variable instead of calling `won?` again within the if statement.

Reply:
This seems to put me into an infinite loop when I use it, could be this refactor or the one for the won? method

I'm going to reach out to you on slack to talk a bit about it